File: src/models/crossing/OrderedCrossover.ts
import ICrossing from "./ICrossing";
import Individual from "../Individual";

export default class OrderedCrossover implements ICrossing {
    evaluate(first: Individual, second: Individual): Individual {
        let firstIndex = ~~(Math.random() * first.towns.length);
        let secondIndex = ~~(Math.random() * second.towns.length);

        if (firstIndex > secondIndex)
            [firstIndex, secondIndex] = [secondIndex, firstIndex];

        const firstTowns = first.towns.slice(firstIndex, secondIndex);
        const firstTownsIndices = firstTowns.map(town => town.index);

        const towns = second.towns.filter(
            town => !firstTownsIndices.includes(town.index)
        );

        towns.splice(firstIndex, 0, ...firstTowns);

        return new Individual(towns);
    }
}

File: zad1/Algorithms/Crossing/OrderedCrossover.cs
using System;
using System.Linq;
using zad1.Models;

namespace zad1.Algorithms.Crossing
{
    public class OrderedCrossover : ICrossing
    {
        public Individual Evaluate(Individual first, Individual second)
        {
            var rng = new Random();

            var firstIndex = rng.Next(first.Towns.Count);
            var secondIndex = rng.Next(second.Towns.Count);

            if (firstIndex > secondIndex)
            {
                var temp = firstIndex;
                firstIndex = secondIndex;
                secondIndex = temp;
            }

            var firstTowns = first.Towns.GetRange(firstIndex, secondIndex - firstIndex);
            var firstIndices = firstTowns.Select(town => town.Index).ToHashSet();
            var secondTowns = second.Towns.FindAll(town => !firstIndices.Contains(town.Index));

            var towns = secondTowns.Take(firstIndex).ToList();
            towns.AddRange(firstTowns);
            towns.AddRange(secondTowns.Skip(firstIndex));

            return new Individual(towns);
        }
    }
}